packages/axelar-soroban-std/src/shared_interfaces.rs

```diff
@@ -0,0 +1,398 @@
+use crate::ensure;
+use soroban_sdk::{
+    contractclient, symbol_short, Address, BytesN, ConversionError, Env, FromVal, IntoVal, String,
+    Topics, TryFromVal, Val, Vec,
+};
+
+#[contractclient(name = "OwnershipClient")]
+pub trait OwnershipInterface {
+    fn owner(env: &Env) -> Address;
+}
+
+#[contractclient(name = "UpgradeableClient")]
+pub trait UpgradeableInterface: OwnershipInterface {
+    /// Returns the current version of the contract.
+    fn version(env: &Env) -> String;
+
+    /// Upgrades the contract to a new WASM hash.
+    fn upgrade(env: &Env, new_wasm_hash: BytesN<32>);
+}
+
+pub trait MigratableInterface: UpgradeableInterface {
+    /// Data needed during the migration. Each contract can define its own data type.
+    type MigrationData: FromVal<Env, Val>;
+    /// Error type returned if the migration fails.
+    type Error: Into<soroban_sdk::Error>;
+
+    /// Migrates contract state after upgrading the contract code.
+    fn migrate(env: &Env, migration_data: Self::MigrationData) -> Result<(), Self::Error>;
```

Comment:
I'm ok with not returning a result since the caller doesn't do anything meaningful with the error. We can just use `panic_with_error` internally instead to keep this simpler

Reply:
I prefer to keep it as is. Migration can go wrong for a variety of reasons and I don't want to take the choice away from contracts to return a specific error, just because it's likely they will get called from the upgrader contract.
I'm not a fan of how much soroban makes use of panics

Reply:
the upgrader will panic if migrate errors anyways, so doesn't make too much difference until soroban adds error stacking ability. but i'm fine to keep it as-is